enhancements/authentication/improved-login-workflow.md

```diff
@@ -0,0 +1,348 @@
+---
+title: oc-authorization-grant-login
+authors:
+  - "@arjunrn"
+reviewers:
+  - "@stlaz"
+  - "@s-urbaniak"
+approvers:
+  - TBD
+api-approvers: # in case of new or modified APIs or API extensions (CRDs, aggregated apiservers, webhooks, finalizers)
+  - NA
+creation-date: 2021-11-08
+last-updated: 2021-11-08
+status: provisional
+---
+
+# oc Authorization Grant Login
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Operational readiness criteria is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+The current CLI login process involves multiple interactive steps where the user has to enter credentials multiple
+times. This process can be simplified so that the `oc` tool can be used to login directly without having to first login
+to the cluster console, the OAuth server and finally copying the login command. The `oc` tool can be modified to act as
+a public client to the Openshift OAuth server and received an access token
+through [Authorization Grant](https://datatracker.ietf.org/doc/html/rfc6749#section-1.3) flow
+and [PKCE](https://datatracker.ietf.org/doc/html/rfc7636).
+
+## Motivation
+
+Currently, when logging in with the `oc` client a user has to perform the following steps:
+
+1. Open the console in the browser.
+2. Login to console with the configured identity provider.
+3. Click on the **"Copy Login Command"** button which redirects the page to the OAuth application.
+4. Perform login again like in step 2.
+5. Copy the displayed login command.
+6. Run the copied command to configure the client to configure the client and access the cluster.
+
+This process can be simplified to the following steps:
+
+1. User runs `oc login --auto $CLUSTER_API_URL`
```

Comment:
`oc login --code-grant` maybe?

Why would I have to supply $CLUSTER_API_URL?

Reply:
When opening the authorization URL the client(oc) would need the domain and the path. When the path and the domain are not modified it can derive that from the base domain. If it is it needs the extra parameter. Also this is similar to how `oc login --token $TOKEN $CLUSTER_API_URL` works right now.

Reply:
I see, so this is just an optional argument, please have it in `[]`

Reply:
So just to clarify the `$CLUSTER_API_URL` is required to do the first login. On subsequent logins the URL can be retrieved from the existing kubeconfig. Correct?